File: src/test/java/wile/redstonepen/blocks/RedstoneTrackStateTest.java
package wile.redstonepen.blocks;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.stream.IntStream;

import net.minecraft.core.Direction;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;
import org.junit.jupiter.params.provider.MethodSource;
import org.junit.jupiter.params.provider.ValueSource;
import wile.redstonepen.blocks.RedstoneTrack.TrackBlockEntity.TestHooks;
import wile.redstonepen.blocks.RedstoneTrack.defs;
import wile.redstonepen.blocks.RedstoneTrack.defs.connections;

class RedstoneTrackStateTest
{
  private static TestHooks h() { return new TestHooks(); }

  // --- Mask constants ---

  @Test
  void wireMaskCoversExactlyBits0To23()
  {
    assertEquals(0x0000000000ffffffL, defs.STATE_FLAG_WIR_MASK);
  }

  @Test
  void connectionMaskCoversExactlyBits24To29()
  {
    assertEquals(0x000000003f000000L, defs.STATE_FLAG_CON_MASK);
  }

  @Test
  void powerMaskCoversExactlyBits32To55()
  {
    assertEquals(0x00ffffff00000000L, defs.STATE_FLAG_PWR_MASK);
  }

  @Test
  void masksDontOverlap()
  {
    assertEquals(0L, defs.STATE_FLAG_WIR_MASK & defs.STATE_FLAG_CON_MASK);
    assertEquals(0L, defs.STATE_FLAG_WIR_MASK & defs.STATE_FLAG_PWR_MASK);
    assertEquals(0L, defs.STATE_FLAG_CON_MASK & defs.STATE_FLAG_PWR_MASK);
  }

  // --- Wire flags ---

  @Test
  void getWireFlagsZeroStateReturnsZero()
  {
    assertEquals(0, h().getWireFlags());
  }

  @Test
  void getWireFlagsIgnoresUpperBits()
  {
    final TestHooks th = h();
    th.setState(~defs.STATE_FLAG_WIR_MASK); // only non-wire bits set
    assertEquals(0, th.getWireFlags());
  }

  @Test
  void getWireFlagBit0TrueWhenSet()
  {
    final TestHooks th = h();
    th.setState(1L);
    assertTrue(th.getWireFlag(0));
  }

  @Test
  void getWireFlagBit0FalseWhenClear()
  {
    assertFalse(h().getWireFlag(0));
  }

  @Test
  void getWireFlagBit23TrueWhenSet()
  {
    final TestHooks th = h();
    th.setState(1L << 23);
    assertTrue(th.getWireFlag(23));
  }

  @Test
  void getWireFlagCountIs24()
  {
    assertEquals(24, h().getWireFlagCount());
  }

  @Test
  void addWireFlagsDoesNotIncrementForAlreadySetBit()
  {
    final TestHooks th = h();
    th.setState(1L);
    assertEquals(0, th.addWireFlags(1L));
  }

  static IntStream wireBitIndices() { return IntStream.range(0, 24); }

  @ParameterizedTest
  @MethodSource("wireBitIndices")
  void addWireFlagEachBitIndividuallyReturnsOne(int i)
  {
    assertEquals(1, h().addWireFlags(1L << i));
  }

  @Test
  void addWireFlagsAllAtOnceSetsAll24()
  {
    final TestHooks th = h();
    th.addWireFlags(defs.STATE_FLAG_WIR_MASK);
    assertEquals(0x00ffffff, th.getWireFlags());
  }

  // --- Connection flags ---

  @Test
  void getConnectionFlagsZeroStateReturnsZero()
  {
    assertEquals(0, h().getConnectionFlags());
  }

  @Test
  void getConnectionFlagsOnlyBit24SetReturnsOne()
  {
    final TestHooks th = h();
    th.setState(1L << 24);
    assertEquals(1, th.getConnectionFlags());
  }

  @ParameterizedTest
  @ValueSource(ints = {0, 1, 2, 3, 4, 5})
  void getConnectionFlagEachBitIndependently(int i)
  {
    final TestHooks th = h();
    th.setState(1L << (24 + i));
    assertTrue(th.getConnectionFlag(i));
    for(int j = 0; j < 6; ++j) {
      if(j != i) assertFalse(th.getConnectionFlag(j), "flag " + j + " must be clear when only " + i + " is set");
    }
  }

  @Test
  void getConnectionFlagCountIs6()
  {
    assertEquals(6, h().getConnectionFlagCount());
  }

  // --- Side power nibbles ---

  @Test
  void getSidePowerZeroStateReturnsZeroForAllDirections()
  {
    final TestHooks th = h();
    for(Direction dir : Direction.values()) {
      assertEquals(0, th.getSidePower(dir), "getSidePower(" + dir + ") must be 0 on zero state");
    }
  }

  @ParameterizedTest
  @EnumSource(Direction.class)
  void setSidePowerRoundTripsAtMax(Direction dir)
  {
    final TestHooks th = h();
    th.setSidePower(dir, 15);
    assertEquals(15, th.getSidePower(dir));
  }

  @ParameterizedTest
  @EnumSource(Direction.class)
  void setSidePowerRoundTripsAtZero(Direction dir)
  {
    final TestHooks th = h();
    th.setSidePower(dir, 15);
    th.setSidePower(dir, 0);
    assertEquals(0, th.getSidePower(dir));
  }

  @Test
  void setSidePowerDoesNotCorruptAdjacentDirection()
  {
    // Set every direction to 15, then clear each one and verify neighbours unchanged.
    final Direction[] dirs = Direction.values();
    for(int i = 0; i < dirs.length; ++i) {
      final TestHooks th = h();
      for(Direction d : dirs) th.setSidePower(d, 15);
      th.setSidePower(dirs[i], 0);
      for(int j = 0; j < dirs.length; ++j) {
        int expected = (j == i) ? 0 : 15;
        assertEquals(expected, th.getSidePower(dirs[j]),
          "after clearing " + dirs[i] + ", getSidePower(" + dirs[j] + ") wrong");
      }
    }
  }

  @Test
  void setSidePowerTruncatesTo4Bits()
  {
    // 16 = 0b10000 → 4-bit mask → stored as 0
    final TestHooks th = h();
    th.setSidePower(Direction.DOWN, 16);
    assertEquals(0, th.getSidePower(Direction.DOWN));
  }

  // --- getRedstoneDustCount ---

  @Test
  void redstoneDustCountZeroStateReturnsZero()
  {
    assertEquals(0, h().getRedstoneDustCount());
  }

  @Test
  void redstoneDustCountAllWireBitsSetReturns24()
  {
    final TestHooks th = h();
    th.setState(defs.STATE_FLAG_WIR_MASK);
    assertEquals(24, th.getRedstoneDustCount());
  }

  @Test
  void redstoneDustCountAllConnectionBitsSetZeroWireBitsReturns6()
  {
    final TestHooks th = h();
    th.setState(defs.STATE_FLAG_CON_MASK);
    assertEquals(6, th.getRedstoneDustCount());
  }

  @Test
  void redstoneDustCountKWireBitsSetReturnsK()
  {
    for(int k = 0; k <= 24; ++k) {
      final TestHooks th = h();
      th.setState((1L << k) - 1); // lowest k bits set
      assertEquals(k, th.getRedstoneDustCount(), "k=" + k);
    }
  }

  // --- defs.connections static mappings ---

  @Test
  void wireFaceDirectionMappingHas24Entries()
  {
    assertEquals(24, connections.WIRE_FACE_DIRECTION_MAPPING.size() - 1); // key 0L is the zero entry
  }

  @Test
  void wireFaceDirectionMappingKeysAreDistinctPowersOfTwo()
  {
    // Every non-zero key must be a distinct power of two (single bit set).
    final java.util.Set<Long> seen = new java.util.HashSet<>();
    for(Long key : connections.WIRE_FACE_DIRECTION_MAPPING.keySet()) {
      if(key == 0L) continue;
      assertEquals(1, Long.bitCount(key), "key " + Long.toHexString(key) + " must be a single-bit mask");
      assertTrue(seen.add(key), "duplicate key " + Long.toHexString(key));
    }
  }

  @Test
  void bulkFaceMappingHasOneEntryPerFacePlusZero()
  {
    // 6 faces + 1 zero entry = 7
    assertEquals(7, connections.BULK_FACE_MAPPING.size());
  }

  @Test
  void bulkFaceMappingRevNonNullForAllDirections()
  {
    for(Direction dir : Direction.values()) {
      assertNotEquals(null, connections.BULK_FACE_MAPPING_REV.get(dir),
        "BULK_FACE_MAPPING_REV must have entry for " + dir);
    }
  }

  @Test
  void connectionBitOrderRevHasAllSixDirections()
  {
    assertEquals(6, connections.CONNECTION_BIT_ORDER_REV.size());
    for(Direction dir : Direction.values()) {
      assertTrue(connections.CONNECTION_BIT_ORDER_REV.containsKey(dir),
        "CONNECTION_BIT_ORDER_REV missing " + dir);
    }
  }
}

File: common/src/test/kotlin/wile/redstonepen/blocks/RedstoneTrackStateTest.kt
package wile.redstonepen.blocks

import io.kotest.assertions.assertSoftly
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.datatest.withData
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.kotest.property.Arb
import io.kotest.property.arbitrary.int
import io.kotest.property.checkAll
import net.minecraft.core.Direction
import wile.redstonepen.blocks.track.RedstoneTrackDefs.STATE_FLAG_CON_MASK
import wile.redstonepen.blocks.track.RedstoneTrackDefs.STATE_FLAG_PWR_MASK
import wile.redstonepen.blocks.track.RedstoneTrackDefs.STATE_FLAG_WIR_MASK
import wile.redstonepen.blocks.track.RedstoneTrackDefs.connections
import wile.redstonepen.blocks.track.TestHooks

class RedstoneTrackStateTest :
    DescribeSpec({
        fun h() = TestHooks()

        describe("mask constants") {
            it("masks don't overlap") {
                assertSoftly {
                    (STATE_FLAG_WIR_MASK and STATE_FLAG_CON_MASK) shouldBe 0L
                    (STATE_FLAG_WIR_MASK and STATE_FLAG_PWR_MASK) shouldBe 0L
                    (STATE_FLAG_CON_MASK and STATE_FLAG_PWR_MASK) shouldBe 0L
                }
            }
        }

        describe("wire flags") {
            it("zero state returns zero") { h().getWireFlags() shouldBe 0 }

            it("ignores upper bits") {
                val th = h()
                th.state = STATE_FLAG_WIR_MASK.inv()
                th.getWireFlags() shouldBe 0
            }

            it("bit 0 true when set") {
                val th = h()
                th.state = 1L
                th.getWireFlag(0) shouldBe true
            }

            it("bit 0 false when clear") { h().getWireFlag(0) shouldBe false }

            it("bit 23 true when set") {
                val th = h()
                th.state = 1L shl 23
                th.getWireFlag(23) shouldBe true
            }

            it("flag count is 24") { h().getWireFlagCount() shouldBe 24 }

            it("does not increment for already-set bit") {
                val th = h()
                th.state = 1L
                th.addWireFlags(1L) shouldBe 0
            }

            withData((0 until 24).toList()) { i -> h().addWireFlags(1L shl i) shouldBe 1 }

            it("addWireFlags all at once sets all 24") {
                val th = h()
                th.addWireFlags(STATE_FLAG_WIR_MASK)
                th.getWireFlags() shouldBe 0x00ffffff
            }
        }

        describe("connection flags") {
            it("zero state returns zero") { h().getConnectionFlags() shouldBe 0 }

            it("only bit 24 set returns one") {
                val th = h()
                th.state = 1L shl 24
                th.getConnectionFlags() shouldBe 1
            }

            withData(0, 1, 2, 3, 4, 5) { i ->
                val th = h()
                th.state = 1L shl (24 + i)
                th.getConnectionFlag(i) shouldBe true
                for (j in 0 until 6) {
                    if (j != i) th.getConnectionFlag(j) shouldBe false
                }
            }

            it("flag count is 6") { h().getConnectionFlagCount() shouldBe 6 }
        }

        describe("side power") {
            withData(Direction.values().toList()) { dir -> h().getSidePower(dir) shouldBe 0 }

            context("round trips at max") {
                withData(Direction.values().toList()) { dir ->
                    val th = h()
                    th.setSidePower(dir, 15)
                    th.getSidePower(dir) shouldBe 15
                }
            }

            context("round trips at zero after max") {
                withData(Direction.values().toList()) { dir ->
                    val th = h()
                    th.setSidePower(dir, 15)
                    th.setSidePower(dir, 0)
                    th.getSidePower(dir) shouldBe 0
                }
            }

            it("does not corrupt adjacent direction") {
                val dirs = Direction.values()
                for (i in dirs.indices) {
                    val th = h()
                    for (d in dirs) th.setSidePower(d, 15)
                    th.setSidePower(dirs[i], 0)
                    assertSoftly {
                        for (j in dirs.indices) {
                            val expected = if (j == i) 0 else 15
                            th.getSidePower(dirs[j]) shouldBe expected
                        }
                    }
                }
            }

            it("truncates to 4 bits") {
                val th = h()
                th.setSidePower(Direction.DOWN, 16)
                th.getSidePower(Direction.DOWN) shouldBe 0
            }
        }

        describe("dust count") {
            it("zero state returns zero") { h().getRedstoneDustCount() shouldBe 0 }

            it("all wire bits set returns 24") {
                val th = h()
                th.state = STATE_FLAG_WIR_MASK
                th.getRedstoneDustCount() shouldBe 24
            }

            it("all connection bits set, zero wire bits returns 6") {
                val th = h()
                th.state = STATE_FLAG_CON_MASK
                th.getRedstoneDustCount() shouldBe 6
            }

            withData((0..24).toList()) { k ->
                val th = h()
                th.state = (1L shl k) - 1
                th.getRedstoneDustCount() shouldBe k
            }
        }

        describe("side power all values") {
            context("round trips for all values 0..15") {
                withData(Direction.values().toList()) { dir ->
                    checkAll(Arb.int(0..15)) { v ->
                        val th = h()
                        th.setSidePower(dir, v)
                        th.getSidePower(dir) shouldBe v
                    }
                }
            }

            it("all six directions hold distinct values simultaneously") {
                val dirs = Direction.values()
                val th = h()
                dirs.forEachIndexed { i, dir -> th.setSidePower(dir, i + 1) }
                assertSoftly {
                    dirs.forEachIndexed { i, dir -> th.getSidePower(dir) shouldBe i + 1 }
                }
            }
        }

        describe("connection flags by direction") {
            it("CONNECTION_BIT_ORDER index matches getConnectionFlag(index)") {
                assertSoftly {
                    connections.CONNECTION_BIT_ORDER.forEachIndexed { i, dir ->
                        val th = h()
                        th.state = 1L shl (24 + i)
                        th.getConnectionFlag(i) shouldBe true
                        connections.CONNECTION_BIT_ORDER_REV[dir] shouldBe i
                    }
                }
            }
        }

        describe("static mappings") {
            it("WIRE_FACE_DIRECTION_MAPPING has 24 entries") {
                connections.WIRE_FACE_DIRECTION_MAPPING.size - 1 shouldBe 24
            }

            it("WIRE_FACE_DIRECTION_MAPPING keys are distinct powers of two") {
                val seen = mutableSetOf<Long>()
                for (key in connections.WIRE_FACE_DIRECTION_MAPPING.keys) {
                    if (key == 0L) continue
                    java.lang.Long.bitCount(key) shouldBe 1
                    seen.add(key) shouldBe true
                }
            }

            it("BULK_FACE_MAPPING has one entry per face plus zero") {
                connections.BULK_FACE_MAPPING.size shouldBe 7
            }

            it("BULK_FACE_MAPPING_REV non-null for all directions") {
                for (dir in Direction.values()) {
                    connections.BULK_FACE_MAPPING_REV[dir] shouldNotBe null
                }
            }

            it("CONNECTION_BIT_ORDER_REV has all six directions") {
                connections.CONNECTION_BIT_ORDER_REV.size shouldBe 6
                for (dir in Direction.values()) {
                    connections.CONNECTION_BIT_ORDER_REV.containsKey(dir) shouldBe true
                }
            }
        }
    })
